Make unspecified ints to range over negative and positive.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104658

Currently, negative unspecified ints get specialized. This PR creates symbolic values for
unspecified ints (including negative ones).

For example, with this PR, the following code only compiles once, instead of 3 times:

```python
def foo(x, y):
    return torch.fill(torch.zeros(x.shape), y)

foo(10)
foo(-5)
foo(-3)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @anijain2305

ysiraichi committed:
Make unspecified ints to range over negative and positive.

Currently, negative unspecified ints get specialized. This PR creates symbolic values for
unspecified ints (including negative ones).

For example, with this PR, the following code only compiles once, instead of 3 times:

```python
def foo(x, y):
    return torch.fill(torch.zeros(x.shape), y)

foo(10)
foo(-5)
foo(-3)
```

[ghstack-poisoned]

ysiraichi commented on torch/_dynamo/config.py:
```diff
@@ -62,7 +62,7 @@ def is_fbcode():
 # NOTE - this flag can be removed once we can run dynamic_shapes=False w/ the mark_dynamic API
 # see [Note - on the state of mark_dynamic]
-assume_static_by_default = True
+assume_static_by_default = False
```
Temporarily setting this to false so that I can check CI results.

ysiraichi committed:
Fix `reflection_pad2d_backward` lowering. on "Make unspecified ints to range over negative and positive."


Currently, negative unspecified ints get specialized. This PR creates symbolic values for
unspecified ints (including negative ones).

For example, with this PR, the following code only compiles once, instead of 3 times:

```python
def foo(x, y):
    return torch.fill(torch.zeros(x.shape), y)

foo(10)
foo(-5)
foo(-3)
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -982,7 +982,7 @@ def wrap_unspecialized_primitive(self, value):
-                if value < 0 or torch._dynamo.config.specialize_int:
+                if torch._dynamo.config.specialize_int:
                     # Negative values don't create_symbol correctly,
                     # so make sure we do a constant in this case.
```
comment needs updating

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1036,7 +1036,7 @@ def wrap_unspecialized_primitive(self, value):
                     # generate something that is non-negative, but this is
                     # not a sound assumption to make.
                     # Not fixing as this was a preexisting condition.
-                    shape_env.create_symbol(
```
Delete the TODO above

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1045,6 +1045,8 @@ def wrap_unspecialized_primitive(self, value):
                     hint=value,
                     source=self.source,
                 )
+                symbol = wrapped_value.node._expr
```
?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2283,14 +2283,25 @@ def create_unbacked_symbool(self):
         source: Source,
         dynamic_dim: DimDynamic = DimDynamic.DUCK,
         constraint_dim: DimConstraint = None,  # NB: includes None
+        positive: Optional[bool] = True,
```
Is this really Optional[bool]? What does None mean?

ezyang replied:
It seems you can set this to False to mean "never positive", but we don't actually have any concrete use case for this, AFAIK

ysiraichi replied:
That's true. I did this so that I wouldn't need to `positive = positive if positive else None` when I'm actually creating the symbol. Maybe I could just leave a comment when I call it with `positive=None`. What do you think?

ezyang replied:
I think a comment is OK too, just make it hard for people to accidentally say `positive=False` when they meant `positive=None`

ysiraichi replied:
Hmmm. Maybe we could just throw in an assert there like:

```python
def create_symbol(
    ...
    positive: Optional[bool] = True
):
    assert positive != False, (
        "Unspecified integers range over negative and positive numbers. "
        "Setting positive=False means strictly negative. "
        "Use positive=None instead."
    )
```

Let me know what you think.

ysiraichi committed:
Update on "Make unspecified ints to range over negative and positive."


Currently, negative unspecified ints get specialized. This PR creates symbolic values for
unspecified ints (including negative ones).

For example, with this PR, the following code only compiles once, instead of 3 times:

```python
def foo(x, y):
    return torch.fill(torch.zeros(x.shape), y)

foo(10)
foo(-5)
foo(-3)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Make unspecified ints to range over negative and positive."


Currently, negative unspecified ints get specialized. This PR creates symbolic values for
unspecified ints (including negative ones).

For example, with this PR, the following code only compiles once, instead of 3 times:

```python
def foo(x, y):
    return torch.fill(torch.zeros(x.shape), y)

foo(10)
foo(-5)
foo(-3)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 anijain2305

[ghstack-poisoned]

ezyang commented:
By the way, if we notice that this is making codegen worse (because we are unable to assume that something is positive), we could proactively make the ints positive (guarding appropriately) pending seeing a negative value. It didn't seem like it's necessary here but this is something we can put in our back pocket.

ysiraichi commented:
Right. Do you mean: create symbols for positive and negative "on demand"?

ezyang commented:
Essentially, we could proactively assume that a symbol is positive even if we have no evidence for it, and then make it range over negative/positive if we saw a counterexample. Anyway, not for this PR.

ysiraichi commented:
@ezyang @jansel I renamed `create_symint_and_symbol` --> `create_unspecified_symint_and_symbol`, so as to avoid incorrect use. I'm pretty sure I have changed the its only call site. Is it ok to suppress the BC-lint step?

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge
